sql/catalyst/src/main/scala/org/apache/spark/sql/QueryCompilationErrors.scala

```diff
@@ -159,6 +164,174 @@ object QueryCompilationErrors {
+    new AnalysisException(s"ORDER BY position $index is not in select list " +
+      s"(valid range is [1, $size])", t.origin.line, t.origin.startPosition)
+  }
+
+  def groupByPositionRangeError(index: Int, size: Int, t: TreeNode[_]): Throwable = {
+    new AnalysisException(s"GROUP BY position $index is not in select list " +
+      s"(valid range is [1, $size])", t.origin.line, t.origin.startPosition)
+  }
+
+  def generatorNotExpectedError(name: FunctionIdentifier, classCanonicalName: String): Throwable = {
+    new AnalysisException(s"$name is expected to be a generator. However, " +
+      s"its class is $classCanonicalName, which is not a generator.")
+  }
+
+  def distinctOrFilterOnlyWithAggregateFunctionError(prettyName: String): Throwable = {
+    new AnalysisException("DISTINCT or FILTER specified, " +
+      s"but $prettyName is not an aggregate function")
+  }
+
+  def filterInAggregateFunctionInvalidError(): Throwable = {
+    new AnalysisException("FILTER expression is non-deterministic, " +
+      "it cannot be used in aggregate functions")
+  }
+
+  def aliasNumberNotMatchColumnNumberError(
+      columnSize: Int, outputSize: Int, t: TreeNode[_]): Throwable = {
+    new AnalysisException("Number of column aliases does not match number of columns. " +
+      s"Number of column aliases: $columnSize; " +
+      s"number of columns: $outputSize.", t.origin.line, t.origin.startPosition)
+  }
 
+  def distinctOrFilterOnlyWithAggregateFunctionError(
+      aliasesSize: Int, aliasesNames: String): Throwable = {
+    new AnalysisException("The number of aliases supplied in the AS clause does not " +
+      s"match the number of columns output by the UDTF expected $aliasesSize " +
+      s"aliases but got $aliasesNames ")
+  }
+
+  def windowAggregateFunctionWithFilterNotSupportedError(): Throwable = {
+    new AnalysisException("window aggregate function with filter predicate is not supported yet.")
+  }
+
+  def windowFunctionInsideAggregateFunctionNotAllowedError(): Throwable = {
+    new AnalysisException("It is not allowed to use a window function inside an aggregate " +
+      "function. Please use the inner window function in a sub-query.")
+  }
+
+  def expressionWithoutWindowExpressionError(expr: NamedExpression): Throwable = {
+    new AnalysisException(s"$expr does not have any WindowExpression.")
+  }
+
+  def expressionWithMultiWindowExpressionError(
+      expr: NamedExpression, distinctWindowSpec: Seq[WindowSpecDefinition]): Throwable = {
+    new AnalysisException(s"$expr has multiple Window Specifications ($distinctWindowSpec)." +
+      s"Please file a bug report with this error message, stack trace, and the query.")
+  }
 
+  def windowFunctionsNotAllowedInsideWhereClauseError(): Throwable = {
+    new AnalysisException("It is not allowed to use window functions inside WHERE clause")
+  }
+
+  def windowFunctionsNotAllowedInsideHavingClauseError(): Throwable = {
```

Comment:
Maybe we can combine this error and the one above it into one: `windowFunctionNotAllowedError` with the detailed message.

Reply:
I think this work should be put in the ticket for error message optimization.

Reply:
ditto

Reply:
Sounds good.